log/file_test.go

```diff
@@ -0,0 +1,170 @@
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package log
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+)
+
+type bytesBufferWithClose struct {
+	buffer *bytes.Buffer
+}
+
+func (bb *bytesBufferWithClose) Write(p []byte) (int, error) {
+	return bb.buffer.Write(p)
+}
+
+func (bb *bytesBufferWithClose) Close() error {
+	return nil
+}
+
+func Test_FileHookFromConfigLine(t *testing.T) {
+	t.Parallel()
+
+	tests := [...]struct {
+		line       string
+		err        bool
+		errMessage string
+		res        fileHook
+	}{
+		{
+			line: "file",
+			err:  false,
+			res: fileHook{
+				path:   defaultLogfilePath,
+				levels: logrus.AllLevels,
+			},
+		},
+		{
+			line: "file=/tmp/k6.log,level=info",
+			err:  false,
+			res: fileHook{
+				path:   "/tmp/k6.log",
+				levels: logrus.AllLevels[:5],
+			},
+		},
+		{
+			line: "file=./",
+			err:  true,
+		},
+		{
+			line: "file=/a/c/",
+			err:  true,
+		},
+		{
+			line:       "file=,level=info",
+			err:        true,
+			errMessage: "filepath must not be empty",
+		},
+		{
+			line: "file=/tmp/k6.log,level=tea",
+			err:  true,
+		},
+		{
+			line: "file=/tmp/k6.log,unknown",
+			err:  true,
+		},
+		{
+			line: "file=/tmp/k6.log,level=",
+			err:  true,
+		},
+		{
+			line: "file=/tmp/k6.log,level=,",
+			err:  true,
+		},
+		{
+			line:       "file=/tmp/k6.log,unknown=something",
+			err:        true,
+			errMessage: "unknown logfile config key unknown",
+		},
+		{
+			line:       "unknown=something",
+			err:        true,
+			errMessage: "logfile configuration should be in the form `file=path-to-local-file` but is `unknown=something`",
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.line, func(t *testing.T) {
+			t.Parallel()
+
+			res, err := FileHookFromConfigLine(context.Background(), logrus.New(), test.line)
+
+			if test.err {
+				require.Error(t, err)
+
+				if test.errMessage != "" {
+					require.Equal(t, test.errMessage, err.Error())
+				}
+
+				return
+			}
+
+			require.NoError(t, err)
+			test.res.fallbackLogger = res.(*fileHook).fallbackLogger
+			test.res.ch = res.(*fileHook).ch
+			test.res.w = res.(*fileHook).w
+			require.Equal(t, &test.res, res)
+			require.Equal(t, test.res.levels, res.(*fileHook).Levels())
+		})
+	}
+}
+
+func Test_fileHookFire(t *testing.T) {
+	t.Parallel()
+
+	var buff []byte
+	buffer := &bytesBufferWithClose{
+		buffer: bytes.NewBuffer(buff),
+	}
+
+	hook := &fileHook{
+		ch:     make(chan []byte),
+		w:      buffer,
+		levels: logrus.AllLevels,
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(func() {
+		cancel()
+	})
+
+	go hook.loop(ctx)
+
+	logger := logrus.New()
+	logger.AddHook(hook)
+	logger.SetOutput(io.Discard)
+
+	logger.Info("example log line")
+
+	time.Sleep(time.Millisecond * 20)
```

Comment:
```suggestion
cancel()
```
I would prefer to make the concurrency deterministic and not rely on the `time`.

Reply:
We got a race condition here. I consider using `sync.WaitGroup` within a helper struct, something like this
```go
type wgCloser struct {
	io.Writer
	wg sync.WaitGroup
}

func (nc *wgCloser) Close() error {
	nc.wg.Done()
	return nil
}

func TestFileHookFire(t *testing.T) {
	t.Parallel()

	var buffer bytes.Buffer
	wc := &wgCloser{
		Writer: &buffer,
		wg:     sync.WaitGroup{},
	}

	hook := &fileHook{
		ch:     make(chan []byte),
		w:      wc,
		levels: logrus.AllLevels,
	}

	ctx, cancel := context.WithCancel(context.Background())

	wc.wg.Add(1)
	go hook.loop(ctx)

	logger := logrus.New()
	logger.AddHook(hook)
	logger.SetOutput(io.Discard)

	logger.WithContext(ctx).Info("example log line")

	cancel()
	wc.wg.Wait()

	assert.Contains(t, buffer.String(), "example log line")
}
```
What do you think?

Reply:
waiting for channel closing could simplify a bit the code

```go
cancel()
<-hook.ch // wait for ch closing
```